File: Python3/Hard/WrongAnswer/non-negative-integers-without-consecutive-ones_2.py
class Solution:
    def findIntegers(self, n: int) -> int:
        
        # dp[i][0] = dp[i-1][0]+dp[i-1][1]
        # dp[i][1] = dp[i-1][0]
        dp = [[0,0] for _ in range(32)]
        dp[0][0] = 1
        for i in range(1,32):
            dp[i][0] = dp[i-1][0]+dp[i-1][1]
            dp[i][1] = dp[i-1][0]
        res = 0
        pre = 0
        for i in range(31,-1,-1):
            if n&(1<<i):
                res += dp[i][0]+dp[i][1]
                if pre == 1:
                    res -= dp[i][1]
                    break
                pre = 1
            else:
                pre = 0
        return res+1

File: Rust/Hard/WrongAnswer/non-negative-integers-without-consecutive-ones_2.rs
impl Solution {
    pub fn find_integers(n: i32) -> i32 {
        
        let mut n = n;
        let mut ans = 0;
        let mut a = 0;
        let mut b = 1;
        while n > 0 {
            if n & 1 > 0 && n & 2 > 0 {
                ans = a + b;
                a = b;
            } else {
                ans += b;
                a = 0;
            }
            b += a;
            n >>= 1;
        }
        ans
    }
}